Change test-all suite to enable running each test in separate ractors.

luke-gruber opened the pull request:
This tests for ractor safety issues as well as concurrency issues.

You can enable these tests with RUBY_TESTS_WITH_RACTORS=X when running `make test-all TESTS=test/ruby`. Running tests with ractors is currently only working for tests under the "test/ruby" directory. You should also use the `.excludes-ractor` excludes directory.

For example:

```
EXCLUDES=test/.excludes-ractor RUBY_TESTS_WITH_RACTORS=3 TESTS=test/ruby make test-all
```

This will create 3 ractors for each test method, run the test method inside each ractor and then call `join` on the ractors. Then, it's ready to process the next test. The reason we do this instead of taking all the tests and dividing them between the number of ractors is:

* We want to run each test under concurrency load to test whether or not it is safe under ractors. If it isn't safe, we know it's something to do with this specific test.

* If we get a segfault or error, we know the error is coming from this test alone, not interactions with other tests.

The disadvantage of this approach is:

* It's slower than dividing the tests between the number of ractors running.

* This might not uncover ractor scheduling issues that would show up when you have long-running ractors.

ko1 commented on tool/lib/test/unit/testcase.rb:
```diff
@@ -198,22 +215,24 @@ def run runner
       RUN_TEST_TRACE = "#{__FILE__}:#{__LINE__+3}:in `run_test'".freeze
       def run_test(name)
-        progname, $0 = $0, "#{$0}: #{self.class}##{name}"
+        progname, $0 = $0, "#{$0}: #{self.class}##{name}" if main_ractor?
```
TODO:
if we make main Ractor as a manager Ractor (and other Ractors run tests, a manager Ractor knows which tests are running), we can set a running tests on the main Ractor.

ko1 commented on tool/lib/test/unit/testcase.rb:
```diff
@@ -50,6 +50,23 @@ def mingw? platform = RUBY_PLATFORM
+      end
+
+      def multiple_ractors?
+        ENV["RUBY_TESTS_WITH_RACTORS"].to_i > 1
```
cache the value or define different methods with the envval?
some tests can remove ENVVAL.

luke-gruber commented on tool/lib/zombie_hunter.rb:
```diff
@@ -3,7 +3,7 @@
   def after_teardown
     super
-    assert_empty(Process.waitall)
+    assert_empty(Process.waitall) unless multiple_ractors?
```
Since `after_teardown` runs after each test, I didn't want 1 ractor waiting for processes that another had created. I don't think it was causing issues, so I can put it back if you want.

luke-gruber commented on test/test_tmpdir.rb:
```diff
@@ -154,10 +154,10 @@ def test_ractor
             Ractor.receive
           end
         end
-        dir = r.take
```
I'll investigate 🤷

aidenfoxivey reviewed (commented):
Misc. nits + a question.

hsbt commented:
Can you separate tests of `fileutils` and `json`? and submit them to https://github.com/ruby/fileutils and https://github.com/ruby/json?

It seems to work independently from this PR.

luke-gruber committed:
Add more tests

luke-gruber committed:
add another exclude

luke-gruber committed:
Fix a test in test_time.rb

luke-gruber committed:
Timeout.timeout works in ractors now

luke-gruber committed:
Fix multi-ractor tests after rebase

luke-gruber committed:
Change test-all suite to enable running each test in separate ractors.

This tests for ractor safety issues as well as concurrency issues.

You can enable these tests with RUBY_TESTS_WITH_RACTORS=X when running
`make test-all TESTS=test/ruby`. Running tests with ractors is currently only
working for tests under the "test/ruby" directory. You should also use
the `.excludes-ractor` excludes directory.

For example:

```
EXCLUDES=test/.excludes-ractor RUBY_TESTS_WITH_RACTORS=3 make test-all TESTS=test/ruby
```

This will create 3 ractors for each test method, run the test method
inside each ractor and then call `join` on the ractors. Then, it's ready
to process the next test. The reason we do this instead of taking all the
tests and dividing them between the number of ractors is:

* We want to run each test under concurrency load to test whether
  or not it is safe under ractors. If it isn't safe, we know it's
  something to do with this specific test.

* If we get a segfault or error, we know the error is coming from this
  test alone, not interactions with other tests.

The disadvantage of this approach is:

* It's slower than dividing the tests between the number of ractors
  running.

* This might not uncover ractor scheduling issues that would show up when
  you have long-running ractors.
